Microsoft.Azure.Cosmos/src/HttpClient/CosmosHttpClientCore.cs

```diff
@@ -0,0 +1,305 @@
+﻿//------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+//------------------------------------------------------------
+namespace Microsoft.Azure.Cosmos
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Net;
+    using System.Net.Http;
+    using System.Net.Http.Headers;
+    using System.Text;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Cosmos.Resource.CosmosExceptions;
+    using Microsoft.Azure.Documents;
+    using Microsoft.Azure.Documents.Collections;
+
+    internal sealed class CosmosHttpClientCore : CosmosHttpClient
+    {
+        private static readonly TimeSpan GatewayRequestTimeout = TimeSpan.FromSeconds(65);
+        private readonly HttpClient httpClient;
+        private readonly ICommunicationEventSource eventSource;
+
+        private bool disposedValue;
+
+        private CosmosHttpClientCore(
+            HttpClient httpClient,
+            HttpMessageHandler httpMessageHandler,
+            ICommunicationEventSource eventSource)
+        {
+            this.httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
+            this.eventSource = eventSource ?? throw new ArgumentNullException(nameof(eventSource));
+            this.HttpMessageHandler = httpMessageHandler;
+        }
+
+        public override HttpMessageHandler HttpMessageHandler { get; }
+
+        public static CosmosHttpClient CreateWithConnectionPolicy(
+            ApiType apiType,
+            ICommunicationEventSource eventSource,
+            ConnectionPolicy connectionPolicy,
+            HttpMessageHandler httpMessageHandler,
+            EventHandler<SendingRequestEventArgs> sendingRequestEventArgs,
+            EventHandler<ReceivedResponseEventArgs> receivedResponseEventArgs)
+        {
+            if (connectionPolicy == null)
+            {
+                throw new ArgumentNullException(nameof(connectionPolicy));
+            }
+
+            Func<HttpClient> httpClientFactory = connectionPolicy.HttpClientFactory;
+            if (httpClientFactory != null)
+            {
+                if (sendingRequestEventArgs != null &&
+                    receivedResponseEventArgs != null)
+                {
+                    throw new InvalidOperationException($"{nameof(connectionPolicy.HttpClientFactory)} can not be set at the same time as {nameof(sendingRequestEventArgs)} or {nameof(ReceivedResponseEventArgs)}");
+                }
+
+                HttpClient userHttpClient = httpClientFactory?.Invoke() ?? throw new ArgumentNullException(nameof(httpClientFactory));
```

Comment:
nit: (we already know httpClientFactory is not null)

```suggestion
                HttpClient userHttpClient = httpClientFactory.Invoke() ?? throw new ArgumentNullException(nameof(httpClientFactory));
```